design/upload-progress.md

```diff
@@ -0,0 +1,318 @@
+    	    []ResourceIdentifier, error)
+    }
+
+    // BackupItemProgress is an actor that return the progress of an operation on an individual item.
+    type BackupItemProgress interface {
+    	// AppliesTo returns information about which resources this action should be invoked for.
+    	// A BackupItemProgress's Execute function will only be invoked on items that match the returned
+    	// selector. A zero-valued ResourceSelector matches all resources.
+    	AppliesTo() (ResourceSelector, error)
+    
+    	// Execute 
+    	Progress(input *BackupItemProgressInput) (UploadProgress, error)
+    }
+    
+    // BackupItemProgressInput contains the input parameters for the BackupItemProgressInput's Progress function.
+    type BackupItemProgressInput struct {
+    	// Item is the item that was stored in the backup
+    	Item runtime.Unstructured
+    	// SnapshotID is the snapshot ID returned by SnapshotItemAction
+    	SnapshotID string
+    	// Backup is the representation of the restore resource processed by Velero.
+    	Backup *velerov1api.Backup
+    }
+
+
+## Changes in Velero backup format
+
+No changes to the existing format are introduced by this change.  A `<backup-name>-itemsnapshots.json.gz` file will be 
+added that contains the items and snapshot IDs returned by ItemSnapshotAction.  However, the creation of the 
+`velero-backup.json` object will not occur until the backup moves to one of the terminal phases (_Completed_, 
+_PartiallyFailed_, or _Failed_).  Reconciliation should ignore backups that do not have a `velero-backup.json` object.
+
+The cluster that is creating the backup will have the Backup resource present and will be able to manage the backup
+before the backup completes.
+
+If the Backup resource is removed (e.g. Velero is uninstalled) before a backup completes and writes its 
+`velero-backup.json` object, the other objects in the object store for the backup will be effectively orphaned.  This 
+can currently happen but the current window is much smaller.
+
+
+## CSI snapshots
+
+For systems such as EBS, a snapshot is not available until the storage system has transferred the snapshot to
+stable storage.  CSI snapshots expose the _readyToUse_ state that, in the case of EBS, indicates that the snapshot
+has been transferred to durable storage and is ready to be used.  The CSI BackupItemProgress.Progress method will
+poll that field and when completed, return completion.
+
+## vSphere plug-in
+
+The vSphere Plug-in for Velero uploads snapshots to S3 in the background.  This is also a BackupItemAction plug-in,
+it will check the status of the Upload records for the snapshot and return progress.
+
+## Backup workflow changes
+
+The backup workflow remains the same until we get to the point where the `velero-backup.json` object is written.
+At this point, we will queue the backup to a finalization go-routine.  The next backup may then begin.  The finalization
```

Comment:
I am wondering if there is more value in introducing another CRD, say `BackupMonitorRequest`, and have a controller reconcile/update the state of the backup. This makes it more robust to velero server restarts.
WDYT?

Reply:
I think we can get the restart behavior without adding an additional CRD.  Keep it in mind, and when we are implementing if we need it, let's add it then.